File: back-end/src/test/java/pl/straburzynski/campfiresongs/factory/SongFactory.java
package pl.straburzynski.campfiresongs.factory;

import pl.straburzynski.campfiresongs.song.model.Song;
import pl.straburzynski.campfiresongs.song.model.SongDto;

import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class SongFactory {

    public static Song createSong(UUID id) {
        return Song.builder()
                .id(id)
                .author("author")
                .title("title")
                .lyrics("lyrics")
                .build();
    }

    public static SongDto createSongDto(UUID id) {
        return SongDto.builder()
                .id(id)
                .author("author")
                .title("title")
                .lyrics("lyrics")
                .build();
    }

    public static List<SongDto> createSongDtoList(Integer songsNumber) {
        return IntStream.range(0, songsNumber)
                .mapToObj(i -> createSongDto(UUID.randomUUID()))
                .collect(Collectors.toList());
    }

    public static String createSongData(UUID id) {
        return "{\"id\":" + id + ",\"author\":\"author\",\"title\":\"title\",\"lyrics\":\"lyrics\"}";
    }
}

File: back-end-kotlin/src/test/kotlin/pl/straburzynski/campfiresongs/utils/SongFactory.kt
package pl.straburzynski.campfiresongs.utils

import org.springframework.http.MediaType
import org.springframework.test.web.reactive.server.WebTestClient
import org.springframework.test.web.reactive.server.returnResult
import org.springframework.web.reactive.function.BodyInserters
import pl.straburzynski.campfiresongs.song.model.SongDto
import pl.straburzynski.campfiresongs.utils.Utils.randomString
import java.util.UUID

object SongFactory {
    fun createSong(
        client: WebTestClient,
        requestBody: String = createSongPayload()
    ): SongDto = checkNotNull(
        client.post()
            .uri("/songs")
            .contentType(MediaType.APPLICATION_JSON)
            .body(BodyInserters.fromValue(requestBody))
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .returnResult<SongDto>()
            .responseBody
            .blockFirst()
    )

    fun createMultipleSongs(client: WebTestClient, count: Int): List<SongDto> =
        (1..count).map { _ -> createSong(client, createSongPayload(randomString(5))) }

    fun createSongPayloadWithId(id: UUID, value: String = "test") =
        "{ \"id\": \"$id\",  \"title\": \"$value\", \"author\": \"$value\", \"lyrics\": \"$value\"}"

    fun createSongPayload(value: String = "test"): String =
        "{ \"title\": \"$value\", \"author\": \"$value\", \"lyrics\": \"$value\"}"

    fun buildMultipleSongs(count: Int): List<SongDto> =
        (1..count).map { i ->
            SongDto(
                null,
                "${i}_${randomString(5)}",
                "${i}_${randomString(5)}",
                "${i}_${randomString(5)}"
            )
        }
}

